crates/biome_js_analyze/src/lint/nursery/use_adjacent_overload_signatures.rs

```diff
@@ -0,0 +1,275 @@
+use biome_console::markup;
+use biome_js_syntax::{AnyJsModuleItem, AnyJsStatement, JsModuleItemList, SyntaxNodeText};
+use biome_rowan::{AstNode, TextRange};
+
+declare_rule! {
+    /// Disallow the use of overload signatures that are not next to each other.
+    ///
+    /// Overload signatures must be adjacent.
+    /// If a key is defined multiple times, only the last definition takes effect. Previous definitions are ignored.
+    /// This rule is useful for preventing accidental overloads that are not adjacent.
+    /// It is recommended to keep the overload signatures adjacent to make the code easier to read and maintain.
+    ///
+    /// Source: https://typescript-eslint.io/rules/adjacent-overload-signatures/
+    ///
+    /// ## Examples
+    ///
+    /// ### Invalid
+    ///
+    /// ```js,expect_diagnostic
+    /// type Foo = {
+    ///   foo_type(s: string): void;
+    ///   foo_type(n: number): void;
+    ///   bar_type(): void;
+    ///   foo_type(sn: string | number): void;
+    /// };
+    /// ```
+    ///
+    /// ```js,expect_diagnostic
+    /// interface Foo {
+    ///   foo_interface(s: string): void;
+    ///   foo_interface(n: number): void;
+    ///   bar_interface(): void;
+    ///   foo_interface(sn: string | number): void;
+    /// }
+    /// ``
+    ///
+    /// ```js,expect_diagnostic
+    /// class A {
+    ///   fooA(s: string): void;
+    ///   fooA(n: number): void;
+    ///   barA(): void {};
+    ///   fooA(sn: string | number): void {};
+    /// }
+    /// ```
+    ///
+    /// ### Valid
+    ///
+    /// ```js
+    /// declare namespace Foo {
+    ///   export function foo_declare(s: string): void;
+    ///   export function foo_declare(n: number): void;
+    ///   export function foo_declare(sn: string | number): void;
+    ///   export function bar_declare(): void;
+    /// }
+    /// ````
+    ///
+    /// ```js
+    /// type Foo = {
+    ///   foo_type(s: string): void;
+    ///   foo_type(n: number): void;
+    ///   foo_type(sn: string | number): void;
+    ///   bar_type(): void;
+    /// };
+    /// ```
+    ///
+    /// ```js
+    /// interface Foo {
+    ///   foo_interface(s: string): void;
+    ///   foo_interface(n: number): void;
+    ///   foo_interface(sn: string | number): void;
+    ///   bar_interface(): void;
+    /// }
+    /// ```
+    ///
+    /// ```js
+    /// class A {
+    ///   fooA(s: string): void;
+    ///   fooA(n: number): void;
+    ///   fooA(sn: string | number): void {}
+    ///   barA(): void {}
+    /// }
+    /// ```
+    ///
+    pub UseAdjacentOverloadSignatures {
+        version: "next",
+        name: "useAdjacentOverloadSignatures",
+        language: "js",
+        sources: &[
+            RuleSource::EslintTypeScript("adjacent-overload-signatures")
+        ],
+        recommended: false,
+    }
+}
+
+impl Rule for UseAdjacentOverloadSignatures {
+    type Query = Ast<JsModuleItemList>;
```

Comment:
By querying `JsModuleItemList`, you are not able to handle type/interface and functions defined inside a function. For example:

```ts
function f() {
  interface Inner {
    h(x: boolean): boolean
    g()
    h(x: number): number
  }
}
```

Also, this doesn't handle object type in a type context. For example:

```ts
function f (): {
  h(x: boolean): boolean,
  g(),
  h(x: number): number,
} {
}
```

I think we need to declare a node union using the `declare_node_union!` macro.
We have several rules that use this macro.
This should not be hard to update the current implementation.
Instead of matching over a `ModuleItem`, we will match over the new node union.

Reply:
Oh, I didn't know about declare_node_union!. Thank you for the useful information.

Reply:
Maybe `JsModuleItemList` should be a variant of the node union, because we still want to catch overloads of default export function declarations.

Reply:
> Oh, I didn't know about declare_node_union!. Thank you for the useful information.

Maybe we should update the contribution guide and explain this use case. Sorry about that

Reply:
@ematipico 
np! Yes, that'd be really helpful:)

@Conaclos 
Trying to handle default export function declarations with `declare_node_union!` but it gives below error:

> the trait bound `SyntaxNode<JsLanguage>: From<JsModuleItemList>` is not satisfied

```
declare_node_union! {
    pub DeclarationOsModuleNode = TsTypeAliasDeclaration | TsInterfaceDeclaration | JsClassDeclaration | JsExport | JsFunctionDeclaration | JsModuleItemList // need to handle each `default export function` with `JsExport`
}
```

Reply:
Maybe the trailing comment is the culprit?

Reply:
I don't know why you get this error. Maybe lists cannot be converted to syntax node? Strange...

As a workaround you can use `JsModule | JsScript`. They are the parent of `JsModuleItemList`.

Reply:
@Conaclos 
I did workaround in latest commit:)

Reply:
@chansuke Sis yo push your latest commit? I don't see any change there.

Reply:
@Conaclos 
I pushed my latest update.
I'm using `JsModule` directory.

Reply:
How is the code different from before? Are you giving up on creating a new query node with `declare_node_union`?

Reply:
@Conaclos 
Sorry, I've fixed my code with `declare_node_union`.